### ISSUE TITLE
Update metrics selection panel to initiate creation of custom dimensions

### ISSUE THREAD

**Issue by nfmohit** [google/site-kit-wp#7599]
## Feature Description

The metrics selection panel should be updated so that if a user selects any metric that depends on certain custom dimensions, and if those custom dimensions are not already available, a notification should be added to the footer letting the user know that an update will be made to their Analytics property.

Once they save their selection, if the user doesn't have the Analytics "edit" scope, they should be redirected to the OAuth flow to grant it.

![image](https://github.com/google/site-kit-wp/assets/20284937/67d1c464-26b9-4ceb-93ad-7c4c1b54c389)

See:
1. [Relevant section in the design doc](https://docs.google.com/document/d/1htqStvvyXq_eg5GsHSbdBBgm4Tu_Z6KXsIOBfjWQ5oM/edit?resourcekey=0-Px-yDiCQOTz30DBKllhU3g#heading=h.xv621qkriexs).
2. [Figma designs](https://www.figma.com/file/CdtrHB0HAjfGRWTnTMMQI0/Key-Metrics?type=design&node-id=6570-52606&mode=design&t=jVtKY6g6XWmpvybS-4).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

- If the `newsKeyMetrics` feature flag is on:
  -  In the metrics selection panel, if the user selects any metric that depends on any custom dimensions that aren't yet available in the metric selection panel:
      - A notification should be added according to the Figma designs to the footer of the selection panel.
        - The notification text should vary depending on whether the user has the Analytics “**_edit_**” scope or not.
          - If they do not have the necessary scope, the text should be: `The metrics you selected require more data tracking. You will be directed to update your Analytics property after saving your selection.`
          - If they have the necessary scope, the text should be: `The metrics you selected require more data tracking. We will update your Analytics property after saving your selection.`
      - If the user doesn't have the Analytics "**_edit_**" scope, upon clicking on the "Save Selection" button in the metric selection panel, after the selection is saved, they should be directed to the OAuth flow to grant it in such a way so that Site Kit automatically creates the custom dimensions after they return.
  - Upon closing the selection panel or returning to the Site Kit dashboard from OAuth, if the user already has the Analytics "_**edit**_" scope, the Key Metrics Widget area should create the custom dimensions automatically.
  - The loading state and error handling will be handled per metric in [#7601](https://github.com/google/site-kit-wp/issues/7601). 
  - If the current user is a view-only user, the metrics selection panel shouldn't show any metric for selection that depends on one or more custom dimensions that are not available.

## Implementation Brief

- In `assets/js/modules/analytics-4/datastore/constants.js`:
  - For a new `CORE_FORMS` form ID, create a new constant named `FORM_CUSTOM_DIMENSIONS_CREATE` with the value `analytics4CustomDimensionsCreate`.
- In `assets/js/components/KeyMetrics/MetricsSelectionPanel/Footer.js`:
  - If the `newsKeyMetrics` feature flag is on:
    - Check if any of the metrics the user selects depend on custom dimensions.
      - This can be done by checking the definition for the key metric widget tile within the `KEY_METRICS_WIDGETS` constant (`assets/js/components/KeyMetrics/key-metrics-widgets.js`) under a new `requiredCustomDimensions` array property. This array property will contain names of custom dimensions that the tile depends on. (This will be implemented in #7598)
    - Check if any of the required custom dimensions are not available for the current GA property.
      - This can be done by using the new `hasCustomDimensions` selector (being implemented in [#7598](https://github.com/google/site-kit-wp/issues/7598)).
    - If there are any unavailable custom dimensions, add a notification according to the Figma designs attached to the top of the metric selection footer.
      - The notification should be according to the Figma designs, and should be sticked to the footer so that it doesn't scroll with the metrics.
      - The notification text should be according to the ACs and vary depending on whether the user has the Analytics “**_edit_**” scope or not.
        - The `CORE_USER` `hasScope` selector can be used to determine if the user has the required scope.
    - Upon clicking the footer CTA button, after the metric selection has been saved, do the following if the user doesn't have the Analytics “**_edit_**” scope:
      - Using the `CORE_FORMS` `setValues` action, set the value of an `autoSubmit` property to `true` for the `FORM_CUSTOM_DIMENSIONS_CREATE` form ID.
      - Using the `CORE_USER` `setPermissionScopeError` action, direct the user to the OAuth flow to grant the Analytics “**_edit_**” scope with the modal skipped ([example](https://github.com/google/site-kit-wp/blob/7a4ded2decfd2786ec1ab722e0effce3e59cf36c/assets/js/modules/analytics/components/common/AccountCreate/index.js#L201-L212)).
- In `assets/js/components/DashboardMainApp.js`:
  - On component mount, dispatch the `createCustomDimensions` action (being implemented in [#7598](https://github.com/google/site-kit-wp/issues/7598)) if the following conditions are met:
    - `newsKeyMetrics` feature flag is on.
    - The `CORE_SITE` `isKeyMetricsSetupCompleted` selector returns `true`.
    - The `autoSubmit` property for the `CORE_FORMS` `FORM_CUSTOM_DIMENSIONS_CREATE` form is set to `true`. This can be checked using the `CORE_FORMS` `getValue` selector.
    - The user is authenticated and has the `MANAGE_OPTIONS` capability.
    - The user has the Analytics “**_edit_**” scope.

### Test Coverage

- In `assets/js/components/KeyMetrics/MetricsSelectionPanel/index.test.js`:
  - Add test case(s) to verify the direction of the user to OAuth based on the set conditions.

## QA Brief

* The exact styling of the custom dimension notice will be implemented in a follow up issue. 

#### Authenticated Dashboard, no scope.
* Setup Site Kit on a fresh google account or remove Site Kit Access from [here](https://myaccount.google.com/connections?filters=3), so the `analytics,edit` scope is not available. 
* Setup Site Kit with Analytics, select an already created property, do not enable enhanced measurement. (so that the edit scope is not granted during setup).
* Open Key Metrics Selection Panel and select a tile that needs custom dimension. ie. Least engaging visitor or Pages per visit.
* The notice about custom dimensions (user does not have edit scope) from AC will be displayed.
* Save changes, you should be redirected to 0Auth flow to grand edit scope access.
* After giving access, you'll be returned to dashboard and the required custom dimensions will be created automatically (check Analytics). 
#### Authenticated Dashboard, edit scope.
* Delete the custom dimensions in Analytics. 
* Upon selecting tile that require custom dimension, the notice about custom dimensions (user has edit scope)  will be displayed.
* Apply the changes, the custom dimensions will be created on the fly. 
#### Shared dashboard
* Setup dashboard sharing with analytics access.
* Make sure the property doesn't have the required custom dimensions for KM tiles. 
* On shared dashboard, the aforementioned widgets will not appear in KM selection list. 
* Add the custom dimensions to the property from authenticated dashboard by selecting those tiles and applying. 
* The widget tiles that require custom dimensions will appear in the selection panel list in shared dashboard
* Selecting those tiles will not result any additional notices. 

## Changelog entry

* Allow users to create custom dimensions when selecting certain Key Metric widget tiles.

**Comment by techanvil** [comment]
Hi @nfmohit, thanks for drafting this IB. It's well written, but it does go too much into the implementation detail, reading like a cross between the AC and IB. The AC should ideally be written in a way that non-engineers will be able to read and make sense of it, and if these sort of implementation tips would be useful, they should be included as notes in the IB section (or, potentially as a comment which could be linked to from the IB section).

On a more specific level, it's not entirely clear from reading the AC what the behaviour should be when the user returns from the OAuth flow - should the panel be automatically opened again so the UI returns to the same state, with the creation of the custom dimensions then proceeding automatically?

Lastly, it would be good to specify how we should handle an error from the API call - it looks like we'll need to ensure they are displayed in the same place that save errors are currently shown?

**Comment by nfmohit** [comment]
> Hi @nfmohit, thanks for drafting this IB. It's well written, but it does go too much into the implementation detail, reading like a cross between the AC and IB. The AC should ideally be written in a way that non-engineers will be able to read and make sense of it, and if these sort of implementation tips would be useful, they should be included as notes in the IB section (or, potentially as a comment which could be linked to from the IB section). There are admittedly some issues where it's not possible to write non-implementation-specific AC, but this is one where it should be possible as it's describing UX-centric, user-facing behaviour.

Hi @techanvil. Thank you for the kind review. I've re-written the ACs to be more about the purpose, and moved the previous ACs to the IB section.

> On a more specific level, it's not entirely clear from reading the AC what the behaviour should be when the user returns from the OAuth flow - should the panel be automatically opened again so the UI returns to the same state, with the creation of the custom dimensions then proceeding automatically?

When the user returns from the OAuth flow, the custom dimensions creation will happen in the background (it can be triggered from the Key Metrics Widget area), hence, the panel doesn't need to re-open.

> Lastly, it would be good to specify how we should handle an error from the API call - it looks like we'll need to ensure they are displayed in the same place that save errors are currently shown?

Error handling from the API calls will be done in [#7601](https://github.com/google/site-kit-wp/issues/7601) as we'll be showing errors in individual metrics. As a result, I haven't included it in this issue.

Please let me know if this looks good now. Thanks!

**Comment by techanvil** [comment]
Hey @nfmohit, thanks for the rewrite and for clarifying those points. It's looking very good.

I think it would be useful to include these points of clarification in the AC itself, i.e. point out that upon saving and closing the panel, or upon returning from OAuth, the custom dimensions should be automatically created, with the related tile loading states and any errors from the API calls being displayed via #7601. I'd suggest you can remove the implementation details around `autoSubmit`, `setPermissionScopeError` etc. as these are already captured in the draft IB.

Also, #7601 should probably depend on this issue, what do you think?

**Comment by nfmohit** [comment]
> Hey @nfmohit, thanks for the rewrite and for clarifying those points. It's looking very good.
> 
> I think it would be useful to include these points of clarification in the AC itself, i.e. point out that upon saving and closing the panel, or upon returning from OAuth, the custom dimensions should be automatically created, with the related tile loading states and any errors from the API calls being displayed via #7601. I'd suggest you can remove the implementation details around `autoSubmit`, `setPermissionScopeError` etc. as these are already captured in the draft IB.
> 
> Also, #7601 should probably depend on this issue, what do you think?

Updated, thank you @techanvil !

**Comment by techanvil** [comment]
Thanks @nfmohit, that looks great.

AC :white_check_mark:

**Comment by tofumatt** [comment]
~Overall looks good, but I have one question: where do we get the `requiredCustomDimensions` array from? Is it being added in another issue, or is it the same as what we get from the `hasCustomDimensions` selector (in which case it would be from that selector and not from an array)?~

~I _think_ it's being added in #7598, which is listed as blocking this issue.~

Yup, looks like that's it. I'm just going to note that explicitly in the IB, then this is good 🙂

**Comment by tofumatt** [comment]
IB ✅

**Comment by wpdarren** [comment]
### QA Update: ❌ 

@nfmohit @kuasha420 I have two observations:

1. I set up Site Kit and Analytics and the Google account only has view only set in the Account access management in Analytics, the UX/UI is odd.  I see the Key Metrics widget as expected, select `Pages per visit` metric, and the message appears as per the AC: `The metrics you selected require more data tracking. We will update your Analytics property after saving your selection.` So, the user does not have edit scopes.   I click on the 'Save selection' CTA and are taken to the oAuth page to sign in with my Google account, but this cannot happen because I am set to only view only within Analytics.    It goes through the process, though and I am redirected back to the dashboard with a success notification banner, and the metric does appear. 

In the console though I am seeing errors 403 forbidden which is expected but nothing appears within the UI as an error. 

Is this expected? 🤔  I suspect it's an edge case, but I thought it was worth flagging. 

![image](https://github.com/google/site-kit-wp/assets/73545194/fda03c00-ed2b-4219-819c-40bf6379e2ce)

2,.  I am testing with a Google account that has edit scope to the Analytics account and property.  I have Administrator access to Analytics under Account Access Management for this account.  With this Google account, I set up Site Kit on a fresh new site.  The oAuth came up as expected.  The setup went fine.   I then went to the metrics panel and selected `Pages per visit` metric, and the message that appears is `The metrics you selected require more data tracking. You will be directed to update your Analytics property after saving your selection` instead of `The metrics you selected require more data tracking. We will update your Analytics property after saving your selection.`

The message does not appear as expected when the account has edit scope. 

I did check the custom dimensions on the account, and none appear. 

![image](https://github.com/google/site-kit-wp/assets/73545194/60e85dee-6add-4f9f-b881-1dfbb653bf41)

**Comment by kuasha420** [comment]
Hi @wpdarren, thank you for your observations. To me it looks like you've confused `EDIT_SCOPE` with Edit permissions. The ticket only deals with the scope (granted during oauth flow), the poor UX about resulting from not having edit permission on GA is out of scope for this ticket and an edge case. We can certainly tackle it, but probably in a separate ticket after putting more thoughts on it. 

What do you think, @nfmohit ? 

Cheers.

**Comment by wpdarren** [comment]
@kuasha420  sorry, I have used the wrong terminology in my comments, but the two observations still stand in my opinion.   

1) While view-only permissions in Analytics might be out of scope from the ticket, it is a scenario that could happen where the user doesn't have edit or administrative access on the Analytics level. I could be wrong, but do we not usually display an error on the front end saying that they need to request access?  

2) I have gone through the setup and the oAuth steps to give edit scopes, but the message doesn't change to what's in the AC  I can have a look at this again when I login.

**Comment by nfmohit** [comment]
Hi @wpdarren, thank you for sharing your observations.

> 1. I set up Site Kit and Analytics and the Google account only has view only set in the Account access management in Analytics, the UX/UI is odd.  I see the Key Metrics widget as expected, select `Pages per visit` metric, and the message appears as per the AC: `The metrics you selected require more data tracking. We will update your Analytics property after saving your selection.` So, the user does not have edit scopes.   I click on the 'Save selection' CTA and are taken to the oAuth page to sign in with my Google account, but this cannot happen because I am set to only view only within Analytics.    It goes through the process, though and I am redirected back to the dashboard with a success notification banner, and the metric does appear.
> 
> In the console though I am seeing errors 403 forbidden which is expected but nothing appears within the UI as an error.
> 
> Is this expected? 🤔 I suspect it's an edge case, but I thought it was worth flagging.

This is expected. And you're correct, an error should appear but that is implemented as a part of #7601. Once that is merged, you should see something like this for the affected metric tile(s):

![image](https://github.com/google/site-kit-wp/assets/20284937/9c5688ce-7e10-4569-b818-b733f73c631c)

> 2,. I am testing with a Google account that has edit scope to the Analytics account and property. I have Administrator access to Analytics under Account Access Management for this account. With this Google account, I set up Site Kit on a fresh new site. The oAuth came up as expected. The setup went fine. I then went to the metrics panel and selected `Pages per visit` metric, and the message that appears is `The metrics you selected require more data tracking. You will be directed to update your Analytics property after saving your selection` instead of `The metrics you selected require more data tracking. We will update your Analytics property after saving your selection.`
> 
> The message does not appear as expected when the account has edit scope.

Hmm, just FYI, the initial Site Kit setup doesn't grant the "Analytics edit" scope. It is an additional scope that isn't required by Site Kit to work with the basic features. However, it is an additional scope that is required to create Analytics entities, e.g. custom dimensions, accounts, properties, web data streams, etc.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

- Authenticated Dashboard, no scope.
  - The notice about custom dimensions from AC will be displayed: `The metrics you selected require more data tracking. You will be directed to update your Analytics property after saving your selection.`
  - Once I save changes, I am redirected to 0Auth flow to grand edit scope access. After giving access, I am returned to dashboard and the required custom dimensions are created automatically.

![image](https://github.com/google/site-kit-wp/assets/73545194/bc0ca9c6-cf4b-49fd-be66-863a838b7e7c)

- Authenticated Dashboard, edit scope.
  - Upon selecting tile that requires custom dimension, the notice about custom dimensions is displayed as per AC. `The metrics you selected require more data tracking. We will update your Analytics property after saving your selection.`
  - Applying the changes, the custom dimensions are created automatically.

![image](https://github.com/google/site-kit-wp/assets/73545194/62afc6fd-cc8b-464d-ac39-cd876198eb42)

- Shared dashboard
  - On the shared dashboard, the widgets do not appear in KM selection list.
  - When the custom dimensions to the property from authenticated dashboard are actioned, the custom dimensions appear in the selection panel list in the shared dashboard
  - Selecting those tiles do not result any additional notices.

![image](https://github.com/google/site-kit-wp/assets/73545194/ed78bdab-971e-41a2-ba59-6572f2f120c0)
![image](https://github.com/google/site-kit-wp/assets/73545194/f8a7a774-c1b5-42a2-a4d7-b6af857da759)
